services/spar/src/Spar/DataMigration/V1_ExternalIds.hs

```diff
@@ -0,0 +1,196 @@
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Spar.DataMigration.V1_ExternalIds where
+
+import Cassandra
+import qualified Cassandra as C
+import Control.Lens
+import Data.Conduit
+import qualified Data.Conduit.Combinators as CC
+import Data.Conduit.Internal (zipSources)
+import qualified Data.Conduit.List as CL
+import Data.Id
+import qualified Data.Map.Strict as Map
+import Imports
+import Spar.DataMigration.RIO (RIO (..), modifyRef, readRef, runRIO)
+import Spar.DataMigration.Types hiding (logger)
+import qualified Spar.DataMigration.Types as Types
+import System.Logger (Logger)
+import qualified System.Logger as Log
+
+migration :: Migration
+migration =
+  Migration
+    { version = MigrationVersion 1,
+      text = "Backfill spar.scim_external",
+      action = migrationAction
+    }
+
+type LegacyExternalId = (Text, UserId)
+
+type UserTeam = (UserId, Maybe TeamId)
+
+type NewExternalId = (TeamId, Text, UserId)
+
+data ResolveTeamResult
+  = UserHasNoTeam UserId Text
+  | NewExternalId NewExternalId
+
+---------------------------------------------------------------------------------
+
+class HasMigEnv env where
+  migEnv :: env -> Env
+
+askMigEnv :: HasMigEnv env => RIO env Env
+askMigEnv = asks migEnv
+
+class HasFailCount env where
+  failCount :: env -> IORef Int32
+
+class HasLogger env where
+  logger :: env -> Logger
+
+logDebug :: HasLogger env => String -> RIO env ()
+logDebug msg = do
+  log_ :: Logger <- asks logger
+  Log.debug log_ (Log.msg msg)
+
+logWarn :: HasLogger env => String -> RIO env ()
+logWarn msg = do
+  log_ :: Logger <- asks logger
+  Log.warn log_ (Log.msg msg)
+
+class HasSpar env where
+  sparClientState :: env -> C.ClientState
+
+runSpar :: HasSpar env => Client a -> RIO env a
+runSpar cl = do
+  cs <- asks sparClientState
+  runClient cs cl
+
+class HasBrig env where
+  brigClientState :: env -> C.ClientState
+
+runBrig :: HasBrig env => Client a -> RIO env a
+runBrig cl = do
+  cs <- asks brigClientState
+  runClient cs cl
+
+---------------------------------------------------------------------------------
+
+data V1Env = V1Env {v1FailCount :: IORef Int32, migrationEnv :: Env}
+
+instance HasSpar V1Env where
+  sparClientState = sparCassandra . migrationEnv
+
+instance HasBrig V1Env where
+  brigClientState = brigCassandra . migrationEnv
+
+instance HasLogger V1Env where
+  logger = Types.logger . migrationEnv
+
+instance HasMigEnv V1Env where
+  migEnv = migrationEnv
+
+instance HasFailCount V1Env where
+  failCount = v1FailCount
+
+migrationAction :: Env -> IO ()
+migrationAction migrationEnv = do
+  v1FailCount <- newIORef 0
+  let v1env = V1Env {..}
+  runRIO v1env migrationMain
+
+migrationMain ::
+  ( HasSpar env,
+    HasBrig env,
+    HasLogger env,
+    HasMigEnv env,
+    HasFailCount env
+  ) =>
+  RIO env ()
+migrationMain = do
```

Comment:
I'm assuming that the rest of this module hasn't changed much, and have just glanced through it.

Reply:
Yes, that module hasn't changed